[SYCL] Add support for __regcall calling convention to spir targets.

kbobrovs opened the pull request:
This calling convention makes compiler return values and function arguments
passed as values (through virtual registers) in most cases. The implementation
is basically customizing SPIRABIInfo for CGFunctionInfo with the X86_RegCall
calling convention. Code generation is mostly borrowed from AMDGPUABIInfo, but
with some of the restrictions for passing by value removed.

Signed-off-by: kbobrovs <Konstantin.S.Bobrovsky@intel.com>

premanandrao commented on clang/test/CodeGenSYCL/regcall-cc-test.cpp:
```diff
@@ -0,0 +1,106 @@
+// RUN:  %clang_cc1 -fsycl-is-device -triple spir64-unknown-unknown-sycldevice -disable-llvm-passes -O0 -emit-llvm %s -o - | FileCheck %s
```
After #3929, we don't need sycldevice in the target triple.

kbobrovs replied:
```suggestion
// RUN:  %clang_cc1 -fsycl-is-device -triple spir64-unknown-unknown -disable-llvm-passes -O0 -emit-llvm %s -o - | FileCheck %s
```

premanandrao reviewed (commented):
I don't have experience with the specifics of the change.   Would request someone else more familiar with calling convention issues to do a more thorough review.

erichkeane commented:
This generally seems ok, though I would probably suggest some additional testing for parameters.

Additionally, I don't think we'd want to accept this without a spec written somewhere (I'd vastly prefer in the clang tree somewhere) that we can compare the implementation to.

kbobrovs commented:
> This generally seems ok, though I would probably suggest some additional testing for parameters.
> 
> Additionally, I don't think we'd want to accept this without a spec written somewhere (I'd vastly prefer in the clang tree somewhere) that we can compare the implementation to.

Thanks! Could you please suggest the test cases? Also, what specific place within clang tree for the doc do you have in mind?

kbobrovs commented on clang/test/CodeGenSYCL/regcall-cc-test.cpp:
```diff
@@ -0,0 +1,106 @@
+// functions. This calling convention makes return values and function arguments
+// passed as values (through virtual registers) in most cases.
+
+// CHECK-DAG: target triple = "spir64-unknown-unknown-sycldevice"
```
```suggestion
// CHECK-DAG: target triple = "spir64-unknown-unknown"
```

erichkeane commented:
> 
> 
> > This generally seems ok, though I would probably suggest some additional testing for parameters.
> > Additionally, I don't think we'd want to accept this without a spec written somewhere (I'd vastly prefer in the clang tree somewhere) that we can compare the implementation to.
> 
> Thanks! Could you please suggest the test cases? Also, what specific place within clang tree for the doc do you have in mind?

/clang/docs is where documentation goes.  There doesn't seem to be a separate file for it, but there is a similar ABI document for Clang Blocks.

As for tests:
-I'd like a few showing that we don't 'register out' like the rest of the ABIs.
-Cases where the same types end up getting used by both parameters and return values (to show we don't care, again, like in register cases).
-Tests that show alignment mismatches don't cause packing.
-Tests with more types.  How are shorts/longs combined?  How about chars and ints? etc.
-At least the documentation has to cover variadic arguments, I believe there is an effort to allow printf, so we'll likely need C variadics covered.

Though, your test names seem very confusing.  It is odd that `PassAsTwo32bitInts` is passed as 8 i8s.  But I expect such an ABI document to cover what you really mean.

kbobrovs commented:
@erichkeane, ok will add more tests. What do you mean by "register out"?

erichkeane commented:
> 
> 
> @erichkeane, ok will add more tests. What do you mean by "register out"?

Typically with ABIs, we change how things are passed based on the number of 'registers' we have used so far.  For example, actual reg-call will use 15 SSE registers and 8 Integer registers (IIRC).  At a certain point we run out of registers, and have to pass on stack.

kbobrovs commented:
> > @erichkeane, ok will add more tests. What do you mean by "register out"?
> 
> Typically with ABIs, we change how things are passed based on the number of 'registers' we have used so far. For example, actual reg-call will use 15 SSE registers and 8 Integer registers (IIRC). At a certain point we run out of registers, and have to pass on stack.

I see. I have a test for that then - AMDGPUABIInfo will register out at 16, I removed that for the SPIRVABIInfo and added a test with 24 parameters - `StillPassThroughRegisters`

kbobrovs commented:
@erichkeane, @AaronBallman - I'm resuming work on this after a big pause, sorry for delay. I addressed your comments - tests extended a lot (one has 130 ints to test for 'register out', various type combinations), also added a description of ABI. I added it to ESIMD, as I think we are not ready to officially support it for pure SYCL.

Please review.

AaronBallman commented on clang/lib/CodeGen/TargetInfo.cpp:
```diff
@@ -10305,17 +10310,114 @@ ABIArgInfo CommonSPIRABIInfo::classifyKernelArgumentType(QualType Ty) const {
+// The two functions below are based on AMDGPUABIInfo, but without any
+// restriction on the maximum number of arguments passed via registers.
+// SPIRV BEs are expected to further adjust the calling convention as
+// needed (use stack or byval-like passing) for some of the arguments.
+
+ABIArgInfo CommonSPIRABIInfo::classifyRegcallReturnType(QualType RetTy) const {
```
```suggestion
// needed (use stack or byval-like passing) for some of the arguments.
ABIArgInfo CommonSPIRABIInfo::classifyRegcallReturnType(QualType RetTy) const {
```

kbobrovs replied:
The comment relates to both functions that follow, not only the `classifyRegcallReturnType`, hence the empty line.

AaronBallman replied:
Sounds reasonable enough to me, thanks!

AaronBallman reviewed (approved):
This is not my strong suit, but I didn't see anything that concerned me.

premanandrao commented on sycl/doc/extensions/experimental/sycl_ext_intel_esimd/sycl_ext_intel_esimd.md:
```diff
@@ -511,6 +511,114 @@ ESIMD_PRIVATE ESIMD_REGISTER(32) simd<int, 16> vc;
+    * 5-8 bytes - array of 2 ints
+  Floating point types are not merged. Structure field alignment rules can
+  increase the calculated size compared to simple sum of `sizeof` of all the
+  types. If the total size exceeds 8, then:
```
> Structure field alignment rules can increase the calculated size compared to simple sum of `sizeof` of all the types.

I think this can be removed since it is mentioned above when describing how size is calculated.

kbobrovs replied:
OK

premanandrao commented on clang/test/CodeGenSYCL/regcall-cc-test.cpp:
```diff
@@ -0,0 +1,360 @@
+// === TEST CASE: multi-level nested structs with one primitive type element at
+// the bottom, and one - at the top. The nested struct at the top is expected to
+// get "unwraped" by the compiler evaporating to the single element at the
+// bottom.
```
I think this comment is a copy-and-paste error describing the previous case.

kbobrovs replied:
oh, right, will fix

premanandrao commented on clang/test/CodeGenSYCL/regcall-cc-test.cpp:
```diff
@@ -0,0 +1,360 @@
+// CHECK-DAG: %struct.CharCharShortFloat = type { i8, i8, i16, float }
+
+template SYCL_DEVICE CharCharShortFloat __regcall func<CharCharShortFloat>(CharCharShortFloat x);
+// CHECK-DAG: define weak_odr x86_regcallcc [2 x i32] @_Z16__regcall3__funcI18CharCharShortFloatET_S1_([2 x i32] %{{[0-9a-zA-Z_.]+}})
```
Could you please explain this?   The doc says floating point types are not merged.

kbobrovs replied:
Thanks for catching! Will update the doc.

premanandrao commented on clang/test/CodeGenSYCL/regcall-cc-test.cpp:
```diff
@@ -0,0 +1,360 @@
+// CHECK-DAG: define weak_odr x86_regcallcc %struct.CharDoubleCharLonglong @_Z16__regcall3__funcI22CharDoubleCharLonglongET_S1_(i8 %{{[0-9a-zA-Z_.]+}}, double %{{[0-9a-zA-Z_.]+}}, i8 %{{[0-9a-zA-Z_.]+}}, i64 %{{[0-9a-zA-Z_.]+}})
+
+
+// === TEST CASE: a struct of 130x4-byte elements is still passed by value
```
Could you add a comment why you chose 130?   Does it hit some threshold over which something different happens?

kbobrovs replied:
This is just "arbitrarily big" number, more than 128 to test big number of fields does not prevent passing by values.
I'll add a comment.

kbobrovs commented:
I'll wait for @erichkeane's review to make review-related changes all at once to reduce stress on the CI system.

erichkeane commented:
I have nothing to add, Aaron and Prem covered everything I believe.

elizabethandrews commented on clang/test/CodeGenSYCL/regcall-cc-test.cpp:
```diff
@@ -0,0 +1,360 @@
+
+// === TEST CASE: multi-level nested structs with one primitive type element at
+// the bottom, and one - at the top. The nested struct at the top is expected to
+// get "unwraped" by the compiler evaporating to the single element at the
```
```suggestion
// get "unwrapped" by the compiler evaporating to the single element at the
```

kbobrovs replied:
thanks for catching!

elizabethandrews reviewed (commented):
Some minor nits.

kbobrovs commented on clang/lib/Basic/Targets/SPIR.h:
```diff
@@ -286,8 +292,10 @@ class LLVM_LIBRARY_VISIBILITY WindowsX86_64_SPIR64TargetInfo
-    if (CC == CC_X86VectorCall)
+    if (CC == CC_X86VectorCall || CC == CC_X86RegCall)
       // Permit CC_X86VectorCall which is used in Microsoft headers
+      // Permit CC_X86RegCall which is used to mark external functions with
+      //   explicit simd or structure type arguments to pass them via registers.
```
ok

kbobrovs commented on clang/lib/Basic/Targets/SPIR.h:
```diff
@@ -128,8 +128,14 @@ class LLVM_LIBRARY_VISIBILITY BaseSPIRTargetInfo : public TargetInfo {
   CallingConvCheckResult checkCallingConvention(CallingConv CC) const override {
-    return (CC == CC_SpirFunction || CC == CC_OpenCLKernel) ? CCCR_OK
-                                                            : CCCR_Warning;
+    return (CC == CC_SpirFunction || CC == CC_OpenCLKernel ||
+            // Permit CC_X86RegCall which is used to mark external functions
+            // with
+            //   explicit simd or structure type arguments to pass them via
+            //   registers.
```
ok

kbobrovs commented on sycl/doc/extensions/experimental/sycl_ext_intel_esimd/sycl_ext_intel_esimd.md:
```diff
@@ -511,6 +511,114 @@ ESIMD_PRIVATE ESIMD_REGISTER(32) simd<int, 16> vc;
+when it (the type) is part of a signature of a function with linkage defined.
+
+Compiler uses aggregate type "unwrapping process" for communicating ARVs.
+Unwarapping a structure with a single field results in the unwrapped type of
```
```suggestion
Unwrapping a structure with a single field results in the unwrapped type of
```

kbobrovs commented:
@intel/dpcpp-esimd-reviewers, friendly ping

kbobrovs commented:
SYCL / Linux / L0 GEN9 LLVM Test Suite (pull_request) failed due to timeouts on unrelated tests:

Timed Out Tests (7):
  SYCL :: KernelAndProgram/multiple-kernel-linking.cpp
  SYCL :: Plugin/level_zero_batch_test.cpp
  SYCL :: Plugin/level_zero_batch_test_copy_with_compute.cpp
  SYCL :: Plugin/level_zero_dynamic_batch_test.cpp
  SYCL :: Plugin/retain_events.cpp
  SYCL :: Printf/char.cpp
  SYCL :: Printf/float.cpp


Testing Time: 1080.59s
  Unsupported      : 184
  Passed           : 578
  Expectedly Failed:  38
  Timed Out        :   7

kbobrovs commented:
@premanandrao, @elizabethandrews, @erichkeane, @AaronBallman - review comments addressed, please take a look/approve

AaronBallman commented on sycl/doc/extensions/experimental/sycl_ext_intel_esimd/sycl_ext_intel_esimd.md:
````diff
@@ -511,6 +511,112 @@ ESIMD_PRIVATE ESIMD_REGISTER(32) simd<int, 16> vc;
+};
+```
+results in { `%struct.B4`, `i32 addrspace(4)*` } pair of types. Thus,
+unwraping can result in a set of a structure, primitive or pointer types -
````
```suggestion
unwrapping can result in a set of a structure, primitive or pointer types -
```

AaronBallman reviewed (approved):
LGTM, though I spotted one more typo in the docs

smanna12 commented on clang/lib/CodeGen/TargetInfo.cpp:
```diff
@@ -10305,17 +10310,114 @@ ABIArgInfo CommonSPIRABIInfo::classifyKernelArgumentType(QualType Ty) const {
+    }
+
+    // Pack aggregates <= 8 bytes into single vector register or pair.
+    // TODO make this parameterizeable/adjustable depending on spir target
```
```suggestion
    // TODO:  Make this parameterizable/adjustable depending on spir target
```

kbobrovs replied:
This part is verbatim copy from AMDGPUABIInfo, so I want to leave it as is

smanna12 replied:
Thanks for the information.

smanna12 commented on clang/lib/CodeGen/TargetInfo.cpp:
```diff
@@ -10305,17 +10310,114 @@ ABIArgInfo CommonSPIRABIInfo::classifyKernelArgumentType(QualType Ty) const {
+      if (Size <= 32)
+        return ABIArgInfo::getDirect(llvm::Type::getInt32Ty(getVMContext()));
+
+      // XXX: Should this be i64 instead, and should the limit increase?
```
```suggestion
      // TODO: Should this be i64 instead, and should the limit increase?
```

kbobrovs replied:
This part is verbatim copy from AMDGPUABIInfo, so I want to leave it as is

smanna12 commented on clang/lib/CodeGen/TargetInfo.cpp:
```diff
@@ -10305,17 +10310,114 @@ ABIArgInfo CommonSPIRABIInfo::classifyKernelArgumentType(QualType Ty) const {
+      }
+
+      // Pack aggregates <= 8 bytes into a single vector register or pair.
+      // TODO make this parameterizeable/adjustable depending on spir target
```
```suggestion
      // TODO: Make this parameterizable/adjustable depending on spir target
```

kbobrovs replied:
This part is verbatim copy from AMDGPUABIInfo, so I want to leave it as is

premanandrao commented on sycl/doc/extensions/experimental/sycl_ext_intel_esimd/sycl_ext_intel_esimd.md:
```diff
@@ -511,6 +511,112 @@ ESIMD_PRIVATE ESIMD_REGISTER(32) simd<int, 16> vc;
+the "unwrapped type set".
+
+- If the unwrapped type set has only primitive types, then compiler will "merge"
+  the resulting types if their total size is less or equal to 8 bytes. The total
```
```suggestion
  the resulting types if their total size is less than or equal to 8 bytes. The total
```

premanandrao commented on sycl/doc/extensions/experimental/sycl_ext_intel_esimd/sycl_ext_intel_esimd.md:
```diff
@@ -511,6 +511,112 @@ ESIMD_PRIVATE ESIMD_REGISTER(32) simd<int, 16> vc;
+- If the unwrapped type set has only primitive types, then compiler will "merge"
+  the resulting types if their total size is less or equal to 8 bytes. The total
+  size is calculated as `sizeof(<top aggregate type>)`, and structure field
+  alignment rules can make it greater then the simple sum of `sizeof` of all
```
```suggestion
  alignment rules can make it greater than the simple sum of `sizeof` of all
```

premanandrao reviewed (dismissed):
LGTM, just a couple of typos in the doc.

v-klochkov reviewed (dismissed):
There are several misprints mentioned by reviewers, should they be fixed before merging?

kbobrovs commented on clang/lib/CodeGen/TargetInfo.cpp:
```diff
@@ -10286,6 +10286,11 @@ class CommonSPIRABIInfo : public DefaultABIInfo {
 
   ABIArgInfo classifyKernelArgumentType(QualType Ty) const;
 
+  // Add new functions rather then overload existing so that these public APIs
```
```suggestion
  // Add new functions rather than overload existing so that these public APIs
```

kbobrovs commented:
misprints fixed

kbobrovs commented:
There is a single failure below. It is unrelated to this change, internal ticket created. The test passes on Windows in all modes, on Linux with OpenCL BE, fails only on Linux with LevelZero BE.
[2022-03-15T05:57:52.148Z] Failed Tests (1):
[2022-03-15T05:57:52.148Z]   SYCL :: Assert/assert_in_simultaneous_kernels.cpp
